gpytorch/random_variables/categorical_random_variable.py

```diff
@@ -1,29 +1,53 @@
 import torch
 from torch.autograd import Variable
-import math
 from .random_variable import RandomVariable
 
 
 class CategoricalRandomVariable(RandomVariable):
     def __init__(self, mass_function):
-        if isinstance(mass_function, Variable):
-            mass_function = mass_function.data
+        """
+        Constructs a categorical random variable
+        mass_function represents the weights of the distribution
+
+        Passing a vector mass function corresponds to a single categorical variable
+        Passing a matrix mass function corresponds to a batch of independent categorial
+            variables.
+
+        Params:
+        - mass_function (Variable: vector k or matrix n x k) weights of categorical distribution
+        """
+        if not isinstance(mass_function, Variable):
+            raise RuntimeError('mass_function should be a Variable')
+
+        ndimension = mass_function.ndimension()
+        if ndimension not in [1, 2]:
+            raise RuntimeError('mass_function should be a vector or a matrix')
+
+        # Assert that probabilities sum to 1
+        if ndimension == 1:
+            mass_function = mass_function.unsqueeze(0)
+        if any(torch.abs(mass_function.data.sum(1) - 1) > 1e-5):
+            raise RuntimeError('mass_function probabilties (in each row) should sum to 1!')
+        if ndimension == 1:
+            mass_function = mass_function.squeeze(0)
 
         self.mass_function = mass_function
 
     def representation(self):
         return self.mass_function
 
-    def log_probability(self, i):
-        if i > len(self.mass_function):
-            raise RuntimeError(' '.join([
-                'Attempted to access a Categorical mass function with a',
-                'category number larger than the total number of categories: %d'.format(i)
-            ]))
-        return math.log(self.mass_function[i])
+    def sample(self, n_samples=1):
+        mass_function = self.mass_function
+        if isinstance(mass_function, Variable):
```

Comment:
In __init__(), we have checked whether `mass_function` is Variable or not.

Reply:
Removed